torch/distributed/fsdp/_symbolic_trace.py

```diff
@@ -0,0 +1,220 @@
+import contextlib
+import functools
+from dataclasses import dataclass, field
+from typing import Any, Callable, Dict, Generator, List, Optional, Set, Tuple
+
+import torch
+
+
+__all__ = ["TracingConfig"]
+
+
+@dataclass
+class TracingConfig:
+    """
+    Configurations used in ``ParamExecOrderWrapPolicy`` for symbolic tracing of a model.
+
+    Args:
+        tracer (torch.fx.Tracer): An instance of ``torch.fx.Tracer`` that will be used to perform symbolic
+            tracing. ``tracer`` is default to be ``torch.fx.Tracer()``, but can also be instance
+            of some child class of ``torch.fx.Tracer``. For example, one may want to use ``HFTracer``
+            for models in Transformers:
+            .. _Transformers: https://huggingface.co/docs/transformers/index
+
+        concrete_args (Optional[Dict[str, Any]]): Concrete arguments that should not be treated
+            as ``torch.fx.Proxy`` when tracing the forward function.
+            ``concrete_args`` allows one to partially specialize the forward function,
+            including removing control flow or data structures.
+            ``concrete_args`` is also the parameter used in :meth:`~torch.fx.Tracer.trace`.
+    """
+
+    tracer: torch.fx.Tracer = torch.fx.Tracer()
+    concrete_args: Optional[Dict[str, Any]] = None
+
+
+@dataclass
+class _ExecutionInfo:
+    """
+    Contains the execution order information in the model forward pass.
+
+    Attributes:
+        current_module: record the module that is currently being traced.
+
+        module_forward_order: a list of modules, where the ordering is based on
```

Comment:
How does `module_forward_order` compare with `module_execution_info_dict`? Are there any invariants we can enforce?

Reply:
synced offline, `module_forward_order` is different from `module_execution_info_dict`. In non-recursive wrapping to be implemented, `module_execution_info_dict` is used as the parameter execution order info and  `module_forward_order` includes the info of how many times a module is called + used to check the forward order in different iterations.